kvserver: decouple cmd checks in replicaAppBatch

tbg opened the pull request:
This refactors the command application pre-flight checks
on replicaAppBatch such that they can move to appBatch
once that struct evolves from a stub into an actual
implementation of `apply.Batch`.

Epic: CRDB-220
Release note: None

tbg committed:
kvserver: decouple cmd checks in replicaAppBatch

This refactors the command application pre-flight checks
on replicaAppBatch such that they can move to appBatch
once that struct evolves from a stub into an actual
implementation of `apply.Batch`.

Epic: CRDB-220
Release note: None

tbg commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -94,39 +95,30 @@ func (b *replicaAppBatch) Stage(
+
+	// TODO(tbg): these assertions should be pushed into
+	// (*appBatch).assertAndCheckCommand.
+	if err := b.assertNoCmdClosedTimestampRegression(ctx, cmd); err != nil {
```
The assertions now run even if forced error is set, but this is fine since then the command is trivial so it shouldn't fail the assertions either.

tbg commented:
Sorry, trying out `git-machete`[^1] and was a bit too trusting that it'd do "right things" throughout.

[^1]: https://github.com/VirtusLab/git-machete

pavelkalinnikov commented on pkg/kv/kvserver/app_batch.go:
```diff
@@ -0,0 +1,76 @@
+	// - changeRemovesReplica
+}
+
+func (b *appBatch) assertAndCheckCommand(
```
Method comments would help. Maybe `appBatch` comment/overview too. I.e. why all these stages? What's the input/output of each stage (or whole sequence of stages) semantically?

tbg replied:
https://github.com/cockroachdb/cockroach/pull/93266 describes the goal:

----

This long (but individually small) sequence of commits moves (*replicaAppBatch).Stage close to the structure that was prototyped in https://github.com/cockroachdb/cockroach/pull/93265, where it has the following steps:

command checks (standalone)
testing interceptors (replica)
pre-add triggers (standalone)
pre-add triggers (replica)
add (to pebble batch, standalone)
post-add triggers (standalone)
post-add triggers (replica)

----

I adopted this comment and put it on `appBatch`. I am planning on expanding the comments significantly in the PR that establishes a functional standalone `appBatch` but let me know pieces that you're confused about today and I'll make sure they get some love earlier.

tbg commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -94,39 +95,30 @@ func (b *replicaAppBatch) Stage(
+
+	var ab appBatch
+
+	leaseIndex, rej, forcedErr, err := ab.assertAndCheckCommand(ctx, &cmd.ReplicatedCmd, &b.state, cmd.IsLocal())
```
Yes, `ab` can only know about `ReplicatedCmd` because `ab` is standalone application. Standalone application must be entirely driven by what's in the log.

tbg commented on pkg/kv/kvserver/replica_application_state_machine.go:
```diff
@@ -79,43 +79,46 @@ func (r *Replica) getStateMachine() *replicaStateMachine {
-	// Consider testing-only filters.
-	if filter := r.store.cfg.TestingKnobs.TestingApplyCalledTwiceFilter; cmd.ForcedErr != nil || filter != nil {
+// TODO(tbg): move this to replica_app_batch.go.
+func replicaApplyTestingFilters(
```
This method shouldn't modify `cmd` - that it does is an oversight, will fix.

I don't want it to modify `cmd` because this code only runs in Replica-level application. I want the testing interceptors to tell me what they will do (the return values from this method) so that we can next pass it to `ab .toCheckedCmd` (which does modify the `cmd`).

tbg committed:
kvserverbase: introduce struct for return val of CheckForcedErr

Epic: CRDB-220
Release note: None

tbg commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -791,8 +783,12 @@ func (mb *ephemeralReplicaAppBatch) Stage(
 
-	mb.r.shouldApplyCommand(ctx, cmd, &mb.state)
-	mb.state.LeaseAppliedIndex = cmd.LeaseIndex
+	leaseIndex, rejection, forcedErr := kvserverbase.CheckForcedErr(
```
Good idea, added.

tbg commented:
Addressed the comments, pushed three new commits, PTAL (Monday is fine!)

tbg committed:
kvserver: fix a testing filter invocation

We were invoking the wrong filter. Luckily, this has only one user, and
for that users, both filters were the same.

Epic: none
Release note: None

tbg commented on pkg/kv/kvserver/replica_application_state_machine.go:
```diff
@@ -79,43 +79,40 @@ func (r *Replica) getStateMachine() *replicaStateMachine {
+			if fr.Rejection == 0 {
+				newFR.Rejection = kvserverbase.ProposalRejectionType(newRej)
 			}
 		} else if feFilter := r.store.cfg.TestingKnobs.TestingApplyForcedErrFilter; feFilter != nil {
```
Oh, well spotted, this is a bug. You are right, we got away with it because `TestReplicaProbeRequest` has `feFilter == filter`. I added a commit to fix this.

tbg commented:
TFTR!

bors r=pavelkalinnikov
